chinese/edit.py

```diff
@@ -29,44 +29,67 @@
 class EditManager:
     def __init__(self):
         addHook('setupEditorButtons', self.setupButton)
-        addHook('loadNote', self.updateButton)
+        addHook('loadNote', self.on_load_note)
         addHook('editFocusLost', self.onFocusLost)
+        gui_hooks.editor_state_did_change.append(self.on_editor_state_did_change)
 
     def setupButton(self, buttons: list[str], editor: aqt.editor.Editor):
-        self.buttonOn = False
-        editor._links['chineseSupport'] = self.onToggle
+        # the button starts as toggled off
+        self.ui_button_toggled_on = False
 
-        button = editor._addButton(
+        button = editor.addButton(
             icon=None,
             cmd='chineseSupport',
+            func=self.onToggle,
             tip='Chinese Support',
             label='<b>汉字</b>',
             id='chineseSupport',
-            toggleable=True)
+            toggleable=True
+        )
 
         return buttons + [button]
 
-    def onToggle(self, editor):
-        self.buttonOn = not self.buttonOn
+    def onToggle(self, editor: aqt.editor.Editor):
+        self.ui_button_toggled_on = not self.ui_button_toggled_on
 
         mid = str(editor.note.note_type()['id'])
 
-        if self.buttonOn and mid not in config['enabledModels']:
+        if self.ui_button_toggled_on and mid not in config['enabledModels']:
             config['enabledModels'].append(mid)
-        elif not self.buttonOn and mid in config['enabledModels']:
+        elif not self.ui_button_toggled_on and mid in config['enabledModels']:
             config['enabledModels'].remove(mid)
 
         config.save()
 
-    def updateButton(self, editor):
-        enabled = str(editor.note.note_type()['id']) in config['enabledModels']
+    def on_editor_state_did_change(
+        self, editor: aqt.editor.Editor, new_state: aqt.editor.EditorState, old_state: aqt.editor.EditorState
+    ):
+        # if the editor just loaded, then we need to set the toggle status of the addon button
+        if old_state is aqt.editor.EditorState.INITIAL:
+            self.updateButton(editor)
+
+    def on_load_note(self, editor: aqt.editor.Editor):
+        # if the editor is still in the initial state then the `NoteEditor` component has not mounted to the DOM yet
+        # meaning that `toggleEditorButton` has not yet been injected and so we can't update the ui button
+        # in this case, we rely on the `editor_state_did_change` to let us know when the editor is ready
+        if editor.state is aqt.editor.EditorState.INITIAL:
+            return
+        self.updateButton(editor)
+
+    def updateButton(self, editor: aqt.editor.Editor):
```

Comment:
I really don't understand how Svelte works but from my naive glance at [NoteEditor.svelte](https://github.com/ankitects/anki/blob/fc24a4be5293530b65d63038a0d0061a76ccedfd/ts/editor/NoteEditor.svelte#L564) this solution seems correct as EditorState is updated and broadcast in the onMount callback. 

Since you reported `loadNote` was being called before the UI is mounted to the DOM (I'm a bit fuzzy on what this actually means though), this code should fix the race condition.

Reply:
Thanks for taking a look. As for the DOM business, in that piece of code you pointed out, since it is in the `onMount`, it runs when the component has been "mounted" aka rendered for the first time. This is problematic for us since even though the browser might be set up, if the component hasn't been rendered yet then the toggle function wouldn't be available for us to use (since it is injected in the `onMount`). Hopefully this clears things up.